PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanelConfigurable.java

```diff
@@ -105,6 +179,137 @@ protected void createUIComponents() {
+            }
+        }
+        // create new module
+        String moduleName = suggestModuleName(LOCAL_ARTIFACT_MODULE_NAME);
+        File moduleFile = new File(myProject.getBasePath(), String.format("%s/%s.iml", moduleName, moduleName));
+        Module module = ModuleManager.getInstance(myProject).newModule(moduleFile.getPath(), ModuleTypeId.JAVA_MODULE);
+        setLocalArtifactModuleName(module.getName());
+        return module;
+    }
+
+    public void addJarToProjectLibrary(@Nullable VirtualFile jarFile) throws IOException {
+        if (jarFile == null) {
+            return;
+        }
+
+        final List<OrderRoot> myRoots = RootDetectionUtil.detectRoots(
+                Arrays.asList(jarFile), null, myProject, new DefaultLibraryRootsComponentDescriptor());
+        final String libraryName = LibraryEditingUtil.suggestNewLibraryName(
+                ProjectLibraryTable.getInstance(myProject).getModifiableModel(), jarFile.getName());
+        final Library library = LibrariesContainerFactory.createContainer(myProject).createLibrary(
+                libraryName, LibrariesContainer.LibraryLevel.PROJECT, myRoots);
+        final Module module = getOrCreateModule();
+        if (module != null) {
+            ModuleRootModificationUtil.addDependency(module, library, DependencyScope.RUNTIME, false);
+
+            setLocalArtifactLibraryName(libraryName);
+            setLocalArtifactPath(submissionPanel.getSelectedArtifactTextFieldText());
+        }
+    }
+
+    public void removeJarFromProjectLibrary() {
+        if (getLocalArtifactLibraryName() == null) {
+            return;
+        }
+
+        final LibraryTable libraryTable = LibraryTablesRegistrar.getInstance().getLibraryTable(myProject);
+        Library library = libraryTable.getModifiableModel().getLibraryByName(getLocalArtifactLibraryName());
+
+        // remove artifact from module dependency
+        final Module module = ModuleManager.getInstance(myProject).findModuleByName(getLocalArtifactModuleName());
+        final ModifiableRootModel rootModel = ModuleRootManager.getInstance(module).getModifiableModel();
+        rootModel.removeOrderEntry(rootModel.findLibraryOrderEntry(library));
+        // remove artifact from library
+        libraryTable.removeLibrary(library);
+
+        setLocalArtifactLibraryName(null);
+        setLocalArtifactPath(null);
+    }
+
+    public void removeModule() throws IOException {
+        if (getLocalArtifactModuleName() == null) {
+            return;
+        }
+
+        final Module module = ModuleManager.getInstance(myProject).findModuleByName(getLocalArtifactModuleName());
+        if (module != null) {
+            VirtualFile moduleDir = module.getModuleFile().getParent();
+            // remove module from project
+            ModuleManager.getInstance(myProject).disposeModule(module);
+            // remove module directory and files
+            moduleDir.delete(this);
```

Comment:
delete `this`?

Reply:
```
  /**
   * Deletes this file. This method should be only called within write-action.
   * See {@link Application#runWriteAction}.
   *
   * @param requestor any object to control who called this method. Note that
   *                  it is considered to be an external change if {@code requestor} is {@code null}.
   *                  See {@link VirtualFileEvent#getRequestor}
   * @throws IOException if file failed to be deleted
   */
  public void delete(Object requestor) throws IOException {
    ApplicationManager.getApplication().assertWriteAccessAllowed();
    LOG.assertTrue(isValid(), "Deleting invalid file");
    getFileSystem().deleteFile(requestor, this);
  }

```

---
In reply to: [209544241](https://github.com/Microsoft/azure-tools-for-java/pull/1848#discussion_r209544241) [](ancestors = 209544241)